pyuvsim/telescope.py

```diff
@@ -37,3 +40,243 @@ def __eq__(self, other):
                 and self.name == other.name
             )
         return False
+
+
+class BeamList(object):
+    """
+    A container for the set of beam models and related parameters.
+
+    Each rank in the simulation gets a copy of the set of beam objects
+    used for calculating Jones matrices. Rather than broadcasting the objects
+    themselves, the list is composed of strings which provide either a complete
+    description of an analytic beam or the path to a beamfits file. After the
+    broadcast, the beams are initialized from these strings.
+
+    This class provides methods to transform between the string and object
+    representations, and also carries parameters used globally by all UVBeams,
+    including their frequency / angular interpolation options.
+
+    This behaves just as a normal Python list in terms of indexing, appending, etc.
+
+    Attributes
+    ----------
+    string_mode : bool
+        Is True if the beams are represented as strings, False if they're objects
+
+    uvb_params : dict
+        Set of additional attributes to set on UVBeam objects.
+
+    Parameters
+    ----------
+    beam_list : list (optional)
+        A list of UVBeam/AnalyticBeam objects or strings describing beams. If
+        beam_list consists of objects, then the BeamList will be initialized in
+        object mode with string_mode == False. If beam_list consists of strings,
+        the BeamList will be initialized in string mode.
+
+        Passing in a mixture of strings and objects will error.
+
+    uvb_params : dict (optional)
+        Options to set uvb_params, overriding settings from passed-in UVBeam objects.
+
+    Raises
+    ------
+    ValueError :
+        For invalid beam_list, or if UVBeams in the input beam_list have
+        conflicting parameters (e.g., one has freq_interp_kind == 'cubic' and
+        another has freq_interp_kind == 'linear').
+
+    Notes
+    -----
+    When a UVBeam is appended to the beam list, its parameters override any in
+    uvb_params.
+
+    If an object beam is added while in string mode, it will be converted to a string
+    before being inserted, and vice versa. ***In order to be converted to a string, a UVBeam
+    needs to have the entry 'beam_path' in its extra_keywords, giving the path to the beamfits
+    file it came from.***
+
+    """
+
+    _float_params = {'sig': 'sigma', 'diam': 'diameter',
+                     'reff': 'ref_freq', 'ind': 'spectral_index'}
+
+    string_mode = True
+
+    def __init__(self, beam_list=None, uvb_params={}):
```

Comment:
I think you don't want to set uvb_params to an empty dict, that can give very strange and unexpected behavior (see e.g. https://docs.python-guide.org/writing/gotchas/). I think you should set it to None and then handle that case explicitly.

Reply:
Huh... I never knew that! Thanks, I'll switch it out.